src/qudi/hardware/fast_adc/spectrum/si_main.py

```diff
@@ -0,0 +1,418 @@
+            self.processor = DataProcessGated(self.data, self.fetcher) if self.ms.gated \
+                else DataProcessorUngated(self.data, self.fetcher)
+            self.commander.init(self.ms, self.processor)
+            self.loop_manager.input_ms(self.ms)
+
+        self.cmd.card.card_reset()
+        self.ms.gated = self.cs.gated
+        self.ms.num_channels = len(self.cs.ai_ch)
+
+        set_dynamic_params(binwidth_s, record_length_s, number_of_gates)
+
+        self.cmd.cfg.configure_all(self.cs)
+        get_buffer_pointer()
+
+        initialize_all()
+
+        return self.ms.binwidth_s, self.ms.record_length_s, self.ms.total_pulse
+
+    def get_status(self):
+        """
+        Receives the current status of the Fast Counter and outputs it as
+                    return value.
+
+                0 = unconfigured
+                1 = idle
+                2 = running
+                3 = paused
+                -1 = error state
+                TODO Use error state when an error is caught.
+        """
+        return self._internal_status
+
+    def start_measure(self):
+        """
+        Start the acquisition and data process loop
+        """
+        self._start_acquisition()
+        self._signal_measure.emit()
+        self.loop_manager.start()
+
+        return 0
+
+    def _start_acquisition(self):
+        """
+        Start the data acquisition
+        """
+        self.log.info('Measurement started')
+        if self._internal_status == CardStatus.idle:
+#            self.loop_manager.init_measure_params()
+            if self.ms.gated:
+                self.cmd.ts_buf.reset_ts_counter()
+                self.cmd.card.start_all_with_extradma()
+            else:
+                self.cmd.card.start_all()
+            self.log.debug('card started')
+
+        elif self._internal_status == CardStatus.paused:
+            self.cmd.card.enable_trigger()
+
+        self.cmd.process.trigger_enabled = True
+        self._internal_status = CardStatus.running
+
+    def get_data_trace(self):
+        """ Polls the current timetrace data from the fast counter.
+
+        Return value is a numpy array (dtype = int64).
+        The binning, specified by calling configure() in forehand, must be
+        taken care of in this hardware class. A possible overflow of the
+        histogram bins must be caught here and taken care of.
+        If the counter is NOT GATED it will return a tuple (1D-numpy-array, info_dict) with
+            returnarray[timebin_index]
+        If the counter is GATED it will return a tuple (2D-numpy-array, info_dict) with
+            returnarray[gate_index, timebin_index]
+
+        info_dict is a dictionary with keys :
+            - 'elapsed_sweeps' : the elapsed number of sweeps
+            - 'elapsed_time' : the elapsed time in seconds
+
+        If the hardware does not support these features, the values should be None
+        """
+        with self._threadlock:
```

Comment:
This thread lock locks sketchy. It is not taken when writing to the same data in DataProcess.